src/vs/workbench/common/backup.ts

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as path from 'path';
+import * as crypto from 'crypto';
+import * as fs from 'original-fs';
+import * as arrays from 'vs/base/common/arrays';
+import Uri from 'vs/base/common/uri';
+import { IWorkspaceContextService } from 'vs/platform/workspace/common/workspace';
+import { IEnvironmentService } from 'vs/platform/environment/common/environment';
+import { IBackupService } from 'vs/platform/backup/common/backup';
+
+interface IBackupFormat {
+	folderWorkspaces?: {
+		[workspacePath: string]: string[]
+	};
+}
+
+export class BackupService implements IBackupService {
+
+	public _serviceBrand: any;
+
+	private workspaceResource: Uri;
+	private workspacesJsonFilePath: string;
+	private fileContent: IBackupFormat;
+
+	constructor(
+		@IEnvironmentService private environmentService: IEnvironmentService,
+		@IWorkspaceContextService contextService: IWorkspaceContextService
+	) {
+		this.workspacesJsonFilePath = path.join(environmentService.userDataPath, 'Backups', 'workspaces.json');
+		this.workspaceResource = contextService.getWorkspace().resource;
+	}
+
+	public getBackupWorkspaces(): string[] {
+		this.load();
+		return Object.keys(this.fileContent.folderWorkspaces || {});
+	}
+
+	public clearBackupWorkspaces(): void {
+		this.fileContent = {
+			folderWorkspaces: {}
+		};
+		this.save();
+	}
+
+	public removeWorkspace(workspace: string): void {
+		this.load();
+		if (!this.fileContent.folderWorkspaces) {
+			return;
+		}
+		delete this.fileContent.folderWorkspaces[workspace];
+		this.save();
+	}
+
+	public getBackupFiles(workspace: string): string[] {
+		this.load();
+		return this.fileContent.folderWorkspaces[workspace] || [];
+	}
+
+	public getBackupUntitledFiles(workspace: string): string[] {
+		const workspaceHash = crypto.createHash('md5').update(this.workspaceResource.fsPath).digest('hex');
```

Comment:
Just fyi we might loosen this restriction in the near future to support more debug scenarios.